core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -323,11 +345,15 @@ private ManifestFile filterManifest(Schema tableSchema, ManifestFile manifest) {
       PartitionSpec spec = reader.spec();
       PartitionAndMetricsEvaluator evaluator =
           new PartitionAndMetricsEvaluator(tableSchema, spec, deleteExpression);
+      boolean hasDeletedFiles = manifestsReferencedForDeletes.contains(manifest.path());
+      if (hasDeletedFiles) {
```

Comment:
I think this structure makes the code harder to reason about and misses an optimization.

This currently checks whether the manifest is known to have a file to remove. If it is unknown, then the logic below is used, which scans the manifest to determine whether it needs to be rewritten. But if the set of referenced manifests is trusted, then there is no need to scan these manifests because the manifest to rewrite for every removal is known.

Reply:
Ah, I see that this may actually be caught by `canContainDeletedFiles`.

I think part of the problem is that this is trying to use instance state rather than passing data through methods. I think the logic would be more clear if state were passed through args rather than relying on instance fields several calls down the chain.

Reply:
Sounds good, I think what we could do is filter manager surfaces the referenced manifests if all the removals are performed with a referenced manifest location. Then in merging snapshot producer we leverage that set to filter down the set that gets passed to the filter manager to begin with

Reply:
Removing the check which led to always rewriting a delete manifest in case it had aged off deletes simplified this, for context see: https://github.com/apache/iceberg/pull/11131#discussion_r1815710103

I moved the referenced check a bit more explicitly up front during the filtering and also avoided some additional referenced set checks (previously we were doing that a few times) but now it should only happen once.